xlators/mgmt/glusterd/src/glusterd-pmap.c

```diff
@@ -152,275 +188,185 @@ pmap_registry_search(xlator_t *this, const char *brickname,
     return 0;
 }
 
-static int
-pmap_registry_search_by_xprt(xlator_t *this, void *xprt,
-                             gf_pmap_port_type_t type)
+int
+pmap_registry_search_by_xprt(xlator_t *this, void *xprt)
 {
     struct pmap_registry *pmap = NULL;
-    int p = 0;
+    struct pmap_ports *tmp_port = NULL;
     int port = 0;
 
     pmap = pmap_registry_get(this);
-
-    for (p = pmap->last_alloc; p >= pmap->base_port; p--) {
-        if (!pmap->ports[p].xprt)
-            continue;
-        if (pmap->ports[p].xprt == xprt) {
-            if (pmap->ports[p].type == type || type == GF_PMAP_PORT_ANY) {
-                port = p;
-                break;
-            }
+    cds_list_for_each_entry(tmp_port, &pmap->ports, port_list)
+    {
+        if (tmp_port->xprt == xprt) {
+            port = tmp_port->port;
+            break;
         }
     }
 
     return port;
 }
 
-static char *
-pmap_registry_search_by_port(xlator_t *this, int port)
-{
-    struct pmap_registry *pmap = NULL;
-    char *brickname = NULL;
-    int max_port = 0;
-
-    max_port = ((glusterd_conf_t *)(this->private))->max_port;
-    if (port > max_port)
-        goto out;
-
-    pmap = pmap_registry_get(this);
-
-    if (pmap->ports[port].type == GF_PMAP_PORT_BRICKSERVER)
-        brickname = pmap->ports[port].brickname;
-
-out:
-    return brickname;
-}
-
 int
-pmap_registry_alloc(xlator_t *this)
+port_brick_bind(xlator_t *this, int port, char *brickname, void *xprt)
 {
     struct pmap_registry *pmap = NULL;
-    int p = 0;
-    int port = 0;
+    struct pmap_ports *tmp_port = NULL;
+    char *tmp_brick;
+    int ret = -1;
 
-    pmap = pmap_registry_get(this);
+    GF_ASSERT(this);
 
-    for (p = pmap->base_port; p <= pmap->max_port; p++) {
-        /* GF_PMAP_PORT_FOREIGN may be freed up ? */
-        if ((pmap->ports[p].type == GF_PMAP_PORT_FREE) ||
-            (pmap->ports[p].type == GF_PMAP_PORT_FOREIGN)) {
-            if (pmap_port_isfree(p)) {
-                pmap->ports[p].type = GF_PMAP_PORT_LEASED;
-                port = p;
-                break;
-            }
+    pmap = pmap_registry_get(this);
+    cds_list_for_each_entry(tmp_port, &pmap->ports, port_list)
```

Comment:
Instead of a list, wouldn't a small hash table be better ?

Reply:
I thought of that, but again that would affect the working of many different parts of code along with brick-mux...so didn't continue with that approach.